### PR TITLE
Multiple Critical Fixes

### PR DESCRIPTION
**Pull request by ebarti**
I tried using your SDK, which did not work at all.

Pagination issues: when a list of users is retrieved, the current implementation **WAS NOT WORKING!**. The last page result set was returned at all times, NOT accomulating the previously retrieved results. Please check the code comments for more details...

Issues identified:
- The result set in users (allData) was not being returned
- Incorrect appending of returned Json values (when result sets are arrays, appending bytes **does not work!**
- Querying Onelogin with a "Cursor" parameter returns a server error as it is invalid. It should start with lowercase.

### CODE REVIEW

File: pkg/services/olhttp/olrequest.go [hunk 1]
```diff
@@ -57,32 +57,33 @@ func (svc OLHTTPService) Read(r interface{}) ([]byte, error) {
...
+		params.Add("cursor", next)
 		req.URL.RawQuery = params.Encode()
 		resp, data, err = svc.executeHTTP(req, resourceRequest)
 	}
 
 	if err != nil {
-		return []byte{}, err
+		return [][]byte{}, err
 	}
-	return data, err
```
Comment (ebarti) [inline pkg/services/olhttp/olrequest.go:85]: This was a very WTF moment for me. This made your tests pass when they should have not.

Comment (ebarti) [reply pkg/services/olhttp/olrequest.go:85]: Also, it hid the fact that the byte data appending you were doing was definitely not correct

File: pkg/services/olhttp/olrequest_test.go [hunk 2]
```diff
@@ -187,13 +187,15 @@ func TestRead(t *testing.T) {
 							return &http.Response{StatusCode: 200, Body: r}, nil
 						}
 
-						resources := queryNResources(2, 0)
-						if req.URL.Query().Get("Cursor") != "" {
-							moreResources := queryNResources(2, 2)
```
Comment (ebarti) [inline pkg/services/olhttp/olrequest_test.go:192]: I have NO idea why you were first returning 2 results, then on cursor advance you were returning 4 results (including the two at the beginning). This probably was done to make tests pass, as your code was definitely not advancing through the pages.

**Comment by dcaponi** [comment]
A couple things:

First of all, I pulled your branch and a bunch of tests fail - please update the mock repository read and re-read methods to return the array of arrays.

After fixing that on my end, I saw that Privileges has a failing test - please fix.

```
--- FAIL: TestGetOne (0.00s)
    --- FAIL: TestGetOne/it_gets_one_privilege_by_id (0.00s)
        privileges_test.go:235: 
                Error Trace:    privileges_test.go:235
```

I'm sorry you were having trouble with this, clearly I am running this with some bad data or something because I have never encountered this problem and you're literally the only person to come forward, so thanks. How did you happen upon this in the first place BTW?

**Comment by ebarti** [comment]
> A couple things:
> 
> First of all, I pulled your branch and a bunch of tests fail - please update the mock repository read and re-read methods to return the array of arrays.
> 
> After fixing that on my end, I saw that Privileges has a failing test - please fix.
> 
> ```
> --- FAIL: TestGetOne (0.00s)
>     --- FAIL: TestGetOne/it_gets_one_privilege_by_id (0.00s)
>         privileges_test.go:235: 
>                 Error Trace:    privileges_test.go:235
> ```
> 
> I'm sorry you were having trouble with this, clearly I am running this with some bad data or something because I have never encountered this problem and you're literally the only person to come forward, so thanks. How did you happen upon this in the first place BTW?

Sorry my bad, I forgot to copy that file in!

So basically I was trying to get a list of users, but I kept getting <1000 users when I knew we have over 15k.

File: pkg/services/apps/v2.go [hunk 3]
```diff
@@ -40,7 +40,15 @@ func (svc *V2Service) Query(query *AppsQuery) ([]App, error) {
...
 
 	var apps []App
-	json.Unmarshal(resp, &apps)
+	for _, bytes := range resp {
+		var unmarshalled []App
+		json.Unmarshal(bytes, &unmarshalled)
+		if len(apps) == 0 {
+			apps = unmarshalled
+		} else {
+			apps = append(apps, unmarshalled...)
```
Comment (ebarti) [inline pkg/services/apps/v2.go:49]: You will see this pattern in multiple v2's: It is related to tests requiring an empty array when the query is valid but there are no matches found.

File: pkg/services/apps/v2.go [hunk 4]
```diff
@@ -58,7 +66,10 @@ func (svc *V2Service) GetOne(id int32) (*App, error) {
 	}
 
 	var app App
-	json.Unmarshal(resp, &app)
+	if len(resp) < 1 {
+		return nil, errors.New("invalid length of response returned")
```
Comment (ebarti) [inline pkg/services/apps/v2.go:70]: Please feel free to modify this if needed

**Review by dcaponi — APPROVED** [review]

**Comment by dcaponi** [comment]
Ok I'll merge this - again, truly sorry about the trouble, and I should have caught this sooner by returning `allData` not `data` and seeing what shakes out of that. I think you're our largest user so nobody's run into this yet which is why it went dormant for so long.

**Comment by dcaponi** [comment]
Look out for v1.1.18

**Pull request merged**